Allow mentors, etc to view diffs in BB format

windymilla opened the pull request:
In order to aid feedback to mentees and similar situations provide an option on the diffs page to view the diffs as BB-code formatted text instead of the usual table view

Sandbox at: https://www.pgdp.org/~windymilla/c.branch/bbdiff

This addresses [Task #1875](https://www.pgdp.net/c/tasks.php?action=show&task_id=1875)

cpeel commented on tools/project_manager/diff.php:
```diff
@@ -138,11 +141,26 @@
 // ---------------------------------------------------------
 
-$diffEngine = new DifferenceEngineWrapper();
+$diffurl = $_SERVER["REQUEST_SCHEME"]."//".$_SERVER['HTTP_HOST'].$_SERVER["PHP_SELF"];
```
Since the HTML will be rendered on this page, we should be able to use the relative URL (just the line below) and not need to reconstruct the full URL. Using relative URLs when we can is preferred. When we can't we use `$code_url` global instead (eg: `$code_url/tools/project_manager/diff.php`).

windymilla replied:
It does need the full URL in this case, since it's not rendering on this page. It's passing the diffurl down into the DiffEngineWrapper and thence to the DiffFormatter. From there, it will be output into the BB-style diffs as `[url=$diffurl]001.png[/url]`, i.e. the URL is included in the PM to the mentee, so the full URL is needed. I've changed it back to use `$code_url` (as I had in my noncvs experiment). It's good to know about the relative URL thing though, thanks.

cpeel replied:
Ah, but the *entry* page is `diff.php` and that's what the link will be submitting to -- the web browser doesn't know about anything else being included. I really think you can get away with a relative URL.

cpeel replied:
Sorry, I didn't read your message clearly. Yup, I see now that we need to include the absolute URL and what you have here is good!

cpeel commented on tools/project_manager/diff.php:
```diff
@@ -138,11 +141,26 @@
+if (user_can_mentor_in_any_round()) {
+    // show option to change diff style
+    if ($bb_diffs) {
+        $bb_diffs_label = _("Table diff style");
```
Should we make this text more action-centric since they're links?
`Show table-style diff` and `Show BBCode-style diff`

cpeel commented on tools/project_manager/diff.php:
```diff
@@ -7,6 +7,7 @@
 include_once($relPath.'links.inc');
 include_once($relPath.'misc.inc'); // get_integer_param(), attr_safe()
 include_once($relPath."DifferenceEngineWrapper.inc");
+include_once($relPath."DifferenceEngineWrapperBB.inc");
```
Your PR is missing at least the `DifferenceEngineWrapperBB.inc` file 😁

windymilla replied:
Oops - thanks.

windymilla replied:
By the way, the `DiffFormatter` include file in `pinc/3rdparty/mediawiki` has a `.php` extension, so originally, I gave my `BBDiffFormatter` which extends the base class a `.php` extension too. One of the security checks then failed, and I guessed it was because `.php` files are checked differently to `.inc`. I therefore renamed my include file to `BBDiffFormatter.inc`. Hope that's OK - if not, what do I need to do to make it pass the security test if it's an include file but with a `.php` extension.

70ray commented:
Since I have never encountered BB-code before, could you say how it would be useful or what difficulties it overcomes?

windymilla commented:
> Since I have never encountered BB-code before, could you say how it would be useful or what difficulties it overcomes?

BBcode is the markup used in DP forum messages. When a mentor (for example) wants to give feedback to a mentee via a PM, they frequently want to include the diffs, e.g. the mentee failed to correct a scanno, and the proofer in the next round has corrected it. The mentor therefore wants to send a message something like

> ----- Line 23, 001.png
> `books arid magazines`
> ...is proofread as...
> `books and magazines`

Copying and pasting the diffs from the standard table of diffs is not easy, and prone to error, and even when the text has been extracted, the mentor still wants to mark up the diffs in mono-spaced font and add a link to the regular diffs page that the mentee can then visit to see it in more detail/context.

So this PR gives an alternative style for the diffs, with the BBCode markup `[pre]` around the text and `[url]` around the link. The link to swap between diff styles will only be visible for users who are allowed to mentor (plus SA/PFs)

BTW, I realised hadn't put a link to [Task #1875](https://www.pgdp.net/c/tasks.php?action=show&task_id=1875) in the PR description at the top, so I'll add that too.

windymilla commented:
I think I've addressed the points raised so far, and it is OK for testing again now. Sandbox https://www.pgdp.org/~windymilla/c.branch/bbdiff has been updated

cpeel commented on pinc/DifferenceEngineWrapperBB.inc:
```diff
@@ -0,0 +1,29 @@
+        if (!$diff && !$otitle) {
+            $header .= "{$ntitle}";
+        } else {
+            $header .= "
+            {$otitle}<br>
+            {$ntitle}";
```
Can we make this and line 27 consistent? Either building the string all in one go and having implicit newlines or doing it piecemeal and being explicit about newlines -- either is fine but the mix is confusing. In general the code is explicit about newlines but we have a mix of building strings piecemeal or doing it all in one go, so whatever you want to do there is fine.

eg:
```php
$header .= "\n$otitle<br>\n$ntitle";
return "$header\n<pre>$diff</pre>\n";
// or
$header .= "\n{$otitle}<br>\n{$ntitle}";
return "{$header}\n<pre>{$diff}</pre>\n";
// or
$header .= "\n" . $otitle . "<br>\n" . $ntitle";
return $header . "\n<pre>" . $diff . "</pre>\n";
```

cpeel commented on pinc/BBDiffFormatter.inc:
```diff
@@ -0,0 +1,74 @@
+                $this->writeOutput("[pre]");
+                $pre_done = true;
+            }
+            $this->writeOutput(html_safe("$prefix${line}"));
```
Either `$prefix$line` or `{$prefix}{$line}` please.

chrismiceli commented on tools/project_manager/diff.php:
```diff
@@ -138,11 +141,25 @@
+    // show option to change diff style
+    if ($bb_diffs) {
+        $bb_diffs_label = _("Show table-style diff");
+        $_GET["bb_diffs"] = "off";
```
is writing to `$_GET` good practice? I assumed it was read only until now.

windymilla replied:
I copied it from the code directly above that sets the "With/Without Formatting" flag. I don't know if it's good practice.

cpeel replied:
It's not read only and it's not a great practice but it's better to be consistent with the other instance in this file (like you have it).

chrismiceli commented on tools/project_manager/diff.php:
```diff
@@ -138,11 +141,25 @@
+        $_GET["bb_diffs"] = "on";
+    }
+    $bb_diffs_url = "?" . attr_safe(http_build_query($_GET));
+    echo "| <a href='$bb_diffs_url'>$bb_diffs_label</a>\n";
```
personally I get super confused when the text I see if for something I'm not currently seeing. For example, when viewing "table-style diff", the button says BB style and vice-versa. Perhaps we could use a dropdown, that shows your current mode so it is clear what you are looking at? Or at least indicate the format in a heading somewhere?

windymilla replied:
The button actually says "Show BBCode-style diff" to make it clear that clicking the link will show BBCode, rather than being a description of what is currently shown. This is the same as the adjacent link, which says "Compare without formatting" when the formatting is being shown. Given that there is currently a checkbox for "Skip empty diffs", and a link for "Compare without formatting", using a dropdown for the BBCode style would be a third method of essentially displaying a toggle switch, all of which are in the same area of the page. 
One thing about the BBCode switch is that it's instantly obvious which style of diffs you are looking at, since they are visually so different. That means it's different to the "without formatting" or the "skip empty diffs" one, where you might not notice which way the flag is set. 
I'm happy to go with the majority though. It might also be worth bearing in mind that the target audience hasn't seen it in the current form yet, so may have some usability suggestions.

chrismiceli reviewed (commented):
Since this code probably is meant to be read, maybe we add a copy to clipboard button to facilitate copying this to be pasted in a forum?

windymilla commented on tools/project_manager/diff.php:
```diff
@@ -138,11 +141,25 @@
 // ---------------------------------------------------------
 
-$diffEngine = new DifferenceEngineWrapper();
+$diffurl = "$code_url/tools/project_manager/diff.php?project=$projectid&amp;image=$image&amp;L_round_num=$L_round_num&amp;R_round_num=$R_round_num";
```
`$image_url` and `$url` in this file use `&amp;` and I followed blindly. If you think the context argument means that this url should be treated differently, I can change it. I'm assuming you mean that I would pass the unencoded version into the constructor, then encode it later when I'm writing it to the webpage?

chrismiceli replied:
My concern is this code doesn't know in what context this string will be used. If it is being output as HTML somewhere, that is where the encoding should be. For all we know at this point, this string may be shoved in a SQL statement, in which case its encoding would need to be different. So either we encode when using it or name it so the encoding is clear to consumers. I think $image_url and $url encode premeptively above because they are used at least within the same file, but even that I would think we should do differently to be honest.

windymilla replied:
Thanks - I hadn't considered that before you raised it

windymilla commented on tools/project_manager/diff.php:
```diff
@@ -169,6 +186,9 @@ function do_navigation($projectid, $image, $L_round_num, $R_round_num, $L_user_c
     $navigation_text .= "\n<input type='hidden' name='L_round_num' value='$L_round_num'>";
     $navigation_text .= "\n<input type='hidden' name='R_round_num' value='$R_round_num'>";
     $navigation_text .= "\n<input type='hidden' name='format' value='$format'>";
+    if ($bb_diffs) {
```
I may be wrong, but does that mean the URLs linked to in the navigation buttons will then always have the `bb_diffs` parameter with its value either on or off? A reason for not doing that is that general visibility of this feature is restricted to mentor-types. I think the way I currently have it means a non-mentor won't see that parameter in the URL when they use a button like `Proofreader Next`. Sorry if I've misunderstood what you are suggesting.

chrismiceli replied:
I wouldn't worry about it being present for non-mentors since it won't have any impact and it will simplify this form code some. Just my opinion, may want to see what others say.

cpeel replied:
I thought about this too, but I think Nigel's approach of not having the param for users who can't see it is the better approach IMHO. Because one of our eagle-eyed users is going to see this param and then try changing it and ask about it, and while it doesn't reveal anything they *can't* see, it's just more hassle than hiding it to begin with.

windymilla commented on tools/project_manager/diff.php:
```diff
@@ -138,11 +141,25 @@
+        $_GET["bb_diffs"] = "on";
+    }
+    $bb_diffs_url = "?" . attr_safe(http_build_query($_GET));
+    echo "| <a href='$bb_diffs_url'>$bb_diffs_label</a>\n";
```
Thanks

chrismiceli commented:
Is it worth it to add php unit tests in any capacity?

windymilla commented:
> Since this code probably is meant to be read, maybe we add a copy to clipboard button to facilitate copying this to be pasted in a forum?

@chrismiceli - did you mean that the text should remain as it is (i.e. with BBCode markup) and a button copy that to the clipboard, or did you mean that the text should be displayed without markup and the copy to clipboard button would add the markup as it copies the text to the clipboard?

That also raises another possibility...scrap pretty much everything I've done, and just add a "Copy diffs to clipboard with BBCode" to the normal diffs page. So, diffs would continue to be shown as they currently are, but when you click the copy to clipboard button it would convert them to the BBCode style as it copies to the clipboard.

windymilla commented:
> Is it worth it to add php unit tests in any capacity?

Possibly - that's still an area I know nothing about

70ray reviewed (commented):
If the lines are very long then a scroll bar appears at the bottom which scrolls the whole window including the header: file:///home/ray/Pictures/Screenshot_20230410_204942.png
Perhaps there should be an enclsing div so only the inside part scrolls.

chrismiceli commented:
> > Since this code probably is meant to be read, maybe we add a copy to clipboard button to facilitate copying this to be pasted in a forum?
> 
> @chrismiceli - did you mean that the text should remain as it is (i.e. with BBCode markup) and a button copy that to the clipboard, or did you mean that the text should be displayed without markup and the copy to clipboard button would add the markup as it copies the text to the clipboard?
> 
> That also raises another possibility...scrap pretty much everything I've done, and just add a "Copy diffs to clipboard with BBCode" to the normal diffs page. So, diffs would continue to be shown as they currently are, but when you click the copy to clipboard button it would convert them to the BBCode style as it copies to the clipboard.

I just meant a button in addition to what you have here. But really we should probably get feedback on intended usage patterns.

70ray commented on pinc/DifferenceEngineWrapperBB.inc:
```diff
@@ -0,0 +1,24 @@
+include_once($relPath."DifferenceEngineWrapper.inc");
+include_once($relPath."BBDiffFormatter.inc");
+
+class DifferenceEngineWrapperBB extends DifferenceEngineWrapper
```
This seems unsymmetrical somehow. Alternatively you could make DifferenceEngineWrapper into an abstract class and extend DifferenceEngineWrapperBB and DifferenceEngineWrapperTable from it. They would have their own constructors and definitions of addHeader() and everything else derived from the parent.

windymilla replied:
Done - it also led me to a way of easily providing the suggested Copy to Clipboard button. Thanks

70ray commented:
> > Since this code probably is meant to be read, maybe we add a copy to clipboard button to facilitate copying this to be pasted in a forum?
> 
> @chrismiceli - did you mean that the text should remain as it is (i.e. with BBCode markup) and a button copy that to the clipboard, or did you mean that the text should be displayed without markup and the copy to clipboard button would add the markup as it copies the text to the clipboard?
> 
> That also raises another possibility...scrap pretty much everything I've done, and just add a "Copy diffs to clipboard with BBCode" to the normal diffs page. So, diffs would continue to be shown as they currently are, but when you click the copy to clipboard button it would convert them to the BBCode style as it copies to the clipboard.

Another possibility would be to make the BBCodestyle look like the processed BBCode in the PM preview and add a copy button as Chris suggested which would copy the source BBCode. I don't know how difficult this would be, perhaps just converting [pre] to <pre> etc. would do it.

70ray commented:
> If the lines are very long then a scroll bar appears at the bottom which scrolls the whole window including the header ...  Perhaps there should be an enclosing div so only the inside part scrolls.

This same effect occurs in other places on the site so probably no need to worry about it.

windymilla commented:
> Another possibility would be to make the BBCodestyle look like the processed BBCode in the PM preview and add a copy button as Chris suggested which would copy the source BBCode. I don't know how difficult this would be, perhaps just converting [pre] to` \<pre\>` etc. would do it.

I haven' t done this for now, though the new class structure easily allows for it. The intended users may have feedback on the visual appearance

chrismiceli commented on tools/project_manager/diff.php:
```diff
@@ -135,17 +138,44 @@
+    $diffEngineCopy = new DifferenceEngineWrapperBB($diffurl);
+    $diffEngineCopy->setText($L_text, $R_text);
+    $cliptext = javascript_safe($diffEngineCopy->getDiff($L_label, $R_label));
+    echo "| \n<input type='button' value='" . attr_safe(_("Copy BBCode to clipboard")) . "' onClick='navigator.clipboard.writeText(\"$cliptext\");'>";
```
`writeText` returns a promise while the user interacts with the permission dialog. If they reject the request or have turned off clipboard access altogether, that promise is rejected. We may want to in that case alert the user the copy failed.
```js
navigator.clipboard.writeText(\"$cliptext\").catch(function () {alert("Unable to copy!");});'
```
but that is getting a bit much. Maybe at this point we inline some js at the top with a click handler and just reference it here in the attribute.

chrismiceli replied:
> @chrismiceli - I added
> 
> ```
>     "js_data" => "
>         function copyToClip(textstring) {
>             navigator.clipboard.writeText(textstring).catch(function () {alert('Unable to copy!');});
>         }
>     ",
> ```
> 
> to `$extra_args` and changed the onClick to
> 
> ```
> onClick='copyToClip(\"$cliptext\");'
> ```
> 
> Is that the kind of thing you had in mind? I am unable to reliably get my browser to reject use of clipboard, but have seen the "Unable to copy" alert. If you are able to control clipboard use easily on your system, this edit is in my [main sandbox](https://www.pgdp.org/~windymilla/c/) but not committed to the bbdiff branch yet if you're able to try it out, or suggest something different. Thanks

That's better. We'd need to encode the $cliptext though for the html attribute context to confirm no characters will escape out of the attribute like ".

windymilla replied:
Doesn't `javascript_safe` already do that to the cliptext before it gets passed in as an argument to `copyToClip`? Or did you mean something else?

chrismiceli replied:
javascript_safe is for encoding things to be used in a javascript context. Html attribute safe if for things going in html attributes. You are doing both here. This is why I was suggesting we pull the diff from the DOM to avoid all encoding issues.

windymilla replied:
Ah yes. Well if the user feedback means we end up being able to use the DOM idea, that will clearly be better. It still might happen that we don't have the button at all. Depending on what others think, the first feedback I got was that they wouldn't want all the diffs in one block in the clipboard, just manually selecting one at a time. We'll see - thanks for continuing to educate me :)

chrismiceli commented on tools/project_manager/diff.php:
```diff
@@ -135,17 +138,44 @@
+if ($bb_diffs) {
+    $diffEngineCopy = new DifferenceEngineWrapperBB($diffurl);
+    $diffEngineCopy->setText($L_text, $R_text);
+    $cliptext = javascript_safe($diffEngineCopy->getDiff($L_label, $R_label));
```
Instead of passing the text directly, maybe just have a function that reads the diff from the DOM so you don't need to encode at all.

chrismiceli replied:
If we're showing the bbcode, I meant grab it from the < pre > tag in the DOM maybe in the function like
```js
navigator.clipboard.writeText(document.getElementById('#preTag').textcontent).catch(function () {
  alert("Unable to copy!");
});
```

windymilla replied:
Thanks for the code snippet - I agree that if we're showing the unmodified bbcode and also not putting the button on the table-style diffs page, that is much neater. From initial user feedback, it may be that the Copy everything to Clipboard button is not really useful for what they do, which is to copy diffs individually into a message template. I'll solicit more feedback after changing/fixing a couple of things, to help make that decision.

cpeel commented on tools/project_manager/diff.php:
```diff
@@ -127,6 +136,7 @@
 
 if ($L_format || $R_format) {
     // show option to change compare method
+    $format_save = array_key_exists("format", $_GET) ? $_GET["format"] : $format;
```
`array_get()` in `misc.inc` is a bit clearer for this:
```php
$format_save = array_get($_GET, "format", $format);
```

cpeel commented on tools/project_manager/diff.php:
```diff
@@ -135,17 +145,49 @@
+
+if (user_can_mentor_in_any_round()) {
+    // show option to change diff style
+    $bb_diffs_save = array_key_exists("bb_diffs", $_GET) ? $_GET["bb_diffs"] : "off";
```
Ditto `array_get()` here.

cpeel commented on tools/project_manager/diff.php:
```diff
@@ -135,17 +145,49 @@
+$diffurl = "$code_url/tools/project_manager/diff.php?project=$projectid&image=$image&L_round_num=$L_round_num&R_round_num=$R_round_num";
+
+// Provide button to copy BBCode diffs not wrapped in HTML
+if ($bb_diffs or true) { // ALWAYS SHOW BUTTON
```
Leftover for testing?

cpeel reviewed (commented):
I'm not 100% sure where this is in the user-testing phase. I'm happy with whatever y'all decide is best from the UX side of things, so let me know when this is ready for a final code review and to go in.

windymilla commented:
> I'm not 100% sure where this is in the user-testing phase. I'm happy with whatever y'all decide is best from the UX side of things, so let me know when this is ready for a final code review and to go in.

Still in the user-testing phase at the moment, thanks - as you saw, I left a couple of things untidy as they may be removed completely or need tidying depending on user feedback. I'll fix the `array_get()` suggestions of course when I do the other tidying.

I'll be sure to let people know when it's ready for final review - thanks

windymilla commented:
For user testing - no need for code review ATM

windymilla commented:
@cpeel & others - thanks for feedback and suggestions. The behavior of the code now appears to match what the users would like, and has had some testing from @srjfoo too. So, it's ready for a code review from whoever is willing - thanks!

srjfoo reviewed (approved):
I'm happy to do more testing as needed. I don't understand this part of the code well enough to be able to give a good code review on it, but as far as it working as expected, it does as far as I can tell.

cpeel commented on pinc/DifferenceEngineWrapper.inc:
```diff
@@ -25,20 +25,28 @@ include_once($relPath."3rdparty/mediawiki/DairikiDiff.php");
 include_once($relPath."3rdparty/mediawiki/WordLevelDiff.php");
 include_once($relPath."3rdparty/mediawiki/DiffEngine.php");
 include_once($relPath."3rdparty/mediawiki/DiffFormatter.php");
-include_once($relPath."3rdparty/mediawiki/TableDiffFormatter.php");
```
Line 21 (pref_options.inc include) needs to be moved into `pinc/DifferenceEngineWrapperTable.inc` as that's where this function is used.

cpeel commented on tools/project_manager/diff.php:
```diff
@@ -135,17 +145,41 @@
 
 $diffEngine->setText($L_text, $R_text);
-$diffEngine->showDiff($L_label, $R_label);
+$difftext = $diffEngine->getDiff($L_label, $R_label);
```
You might consider just echoing here instead of storing it in a variable and then writing it out.

windymilla replied:
Thanks - should be OK now

cpeel reviewed (approved):
Two minor comments and this LGTM!

windymilla commented:
> > I'm happy to do more testing as needed. I don't understand this part of the code well enough to be able to give a good code review on it, but as far as it working as expected, it does as far as I can tell.
> 
> One concern I would have is if the diff had some special characters (html, javascript, &, ", etc.) that caused issues when outputting the HTML or javascript parts of the UI.

Thanks, @chrismiceli. 

To check the characters were being escaped properly, I added the following strings to a page: `&amp;`, `<i>`, `</pre>`. They were reported correctly (as were double quotes) and weren't interpreted as HTML entities/commands. 

See the first diff on [this bage](https://www.pgdp.org/~windymilla/c.branch/bbdiff/tools/project_manager/diff.php?project=projectID567f1ea6a1ba6&image=003.png&L_round_num=4&R_round_num=5&format=keep&jumpto=001.png&only_nonempty_diffs=on&bb_diffs=on). Is that OK?

srjfoo commented:
> > > I'm happy to do more testing as needed. I don't understand this part of the code well enough to be able to give a good code review on it, but as far as it working as expected, it does as far as I can tell.
> > 
> > 
> > One concern I would have is if the diff had some special characters (html, javascript, &, ", etc.) that caused issues when outputting the HTML or javascript parts of the UI.
> 
> Thanks, @chrismiceli.
> 
> To check the characters were being escaped properly, I added the following strings to a page: `&amp;`, `<i>`, `\</pre\>`. They were reported correctly (as were double quotes) and weren't interpreted as HTML entities/commands.
> 
> See the first diff on [this page](https://www.pgdp.org/~windymilla/c.branch/bbdiff/tools/project_manager/diff.php?project=projectID567f1ea6a1ba6&image=003.png&L_round_num=4&R_round_num=5&format=keep&jumpto=001.png&only_nonempty_diffs=on&bb_diffs=on). Is that OK?

And following up on @windymilla's test: see [this page](https://www.pgdp.org/~windymilla/c.branch/bbdiff/tools/project_manager/diff.php?project=projectID4c4a429bed548&image=038.png&L_round_num=4&R_round_num=5&format=keep&jumpto=039.png&bb_diffs=on).
